src/dbnode/storage/index.go

```diff
@@ -475,17 +498,213 @@ func (i *nsIndex) Tick(c context.Cancellable) (namespaceIndexTickResult, error)
 	return result, multiErr.FinalError()
 }
 
-func (i *nsIndex) ReplaceBlock(
-	blockStart time.Time,
-	segments []segment.Segment,
+func (i *nsIndex) Flush(
+	flush persist.IndexFlush,
+	shards []databaseShard,
 ) error {
 	i.state.RLock()
-	defer i.state.RUnlock()
 	if !i.isOpenWithRLock() {
+		i.state.RUnlock()
 		return errDbIndexUnableToFlushClosed
 	}
 
-	return fmt.Errorf("not implemented")
+	flushable := make([]index.Block, 0, len(i.state.blocksByTime))
+	for _, block := range i.state.blocksByTime {
+		if !i.canFlushBlock(block, shards) {
+			continue
+		}
+		flushable = append(flushable, block)
+	}
+	i.state.RUnlock()
+
+	var evictResults index.EvictMutableSegmentResults
+	for _, block := range flushable {
+		immutableSegments, err := i.flushBlock(flush, block, shards)
+		if err != nil {
+			return err
+		}
+		// Make a result that covers the entire time ranges for the
+		// block for each shard
+		fulfilled := make(result.ShardTimeRanges, len(shards))
+		for _, shard := range shards {
+			fulfilled[shard.ID()] = xtime.NewRanges(xtime.Range{
+				Start: block.StartTime(),
+				End:   block.EndTime(),
+			})
```

Comment:
Maybe we should create a helper for given a start/end time and a slice of shards to create a ShardTimeRanges?

Reply:
sure thing

Reply:
done